### ISSUE TITLE
Complex SH?

### ISSUE THREAD

**Issue by Chutlhu** [lab-cosmo/sphericart#125]
Dear all,
Thank you very much for your amazing library.
It works very well, and my collaborators are enthusiastically using it.
We would like to ask if you are thinking of implementing also complex spherical harmonics.

**Comment by cortner** [comment]
It's a rather trivial transformation that should be easy enough to add by a "user". In the Julis code I implemented it here : https://github.com/ACEsuit/Polynomials4ML.jl/blob/main/src/sphericart.jl

**Comment by Chutlhu** [comment]
thank you! I didn't know!
I implemented your solution with a naive for loop, giving the same results as the Scipy implementation.
Thank you so much

**Comment by cortner** [comment]
@ceriottm  -- in general the question is valid. Should complex SH be supported as well? I don't have too strong a view since I will always work with my wrapper package that provides some additional convenience functionality. But I wanted to ask. Feel free to close this again if your group has no desire for the foreseeable future to support this.

**Comment by ceriottm** [comment]
When we started this we meant it to be somewhat "opinionated" - for a while I was thinking of not even providing a normalized version and just compute solid harmonics. It is true that this is a relatively trivial "wrapper" so if @Chutlhu is interested in contributing wrappers for all the languages and frontends we support I'd be welcoming. I'd like to also hear @frostedoyster and @Luthaf though.

**Comment by Chutlhu** [comment]
I could try, but I don't know if I am skilled enough to to this.
I worked on a wrapper for JAX that is differentiable (w.r.t. the input coordinates).
It requires to pre-compute some masks for

**Comment by Luthaf** [comment]
I'm happy to provide complex sph, but following the spirit of #98, we should make sure their API follows the rest of the package.

I'm thinking something like this:


```py
# Torch/Numpy
calculator = ComplexSphericalHarmonics(lmax=10)
sph = calculator(R)

# JAX
calculator = ComplexSphericalHarmonics(lmax=10)
sph = compute_complex_spherical_harmonics(calculator, R)
sph = calculator(R)
```

```julia
# Julia
basis = ComplexSphericalHarmonics(10)

# one of
sph = basis(R)
sph = compute(basis, R)
```

@Chutlhu, do you think the extra masks and indices can be hidden in a "calculator" class?

**Comment by cortner** [comment]
If you decide to do this, then I'll implement the Julia version.

**Comment by Chutlhu** [comment]
Dear @Luthaf, 

> @Chutlhu, do you think the extra masks and indices can be hidden in a "calculator" class?
yes, sure. I updated the [jupyter notebook ](https://colab.research.google.com/drive/1eo09CeWiakTtFFlYHwSfDu8FvJEdqO1z?usp=sharing) with the class for Numpy using the API you suggested.
However, no gradient is currently implemented.
The torch version should be quite similar, but I need more time to set up the testing environment to check that the backpropagation works fine.

Regarding the JAX implementation, it seems that it is currently implemented as a function rather than a class. 
Will this change in future implementations?

**Comment by Luthaf** [comment]
>  Will this change in future implementations?

yes, this is this idea for #98. We don't have a lot of time to work on this currently, but it should happen at some point!

**Comment by ceriottm** [comment]
We (well, @frostedoyster) made an example of wrappers we just merged in #130 . If someone cares enough to implement also forward derivatives and do the same for all frontends, it'll be a quick merge. Meanwhile as soon as I've a minute I'll open a PR to implement #98 , and maybe we can then make a 0.5 release (1.0 is reserved for when we'll have pre-built wheels...)

**Comment by ceriottm** [comment]
I think that with the help available, this is can be considered by having an example of how to compute these with wrappers. I'll close it and if someone thinks there should be an actual separate API we can re-discuss.
